### PR TITLE
[core][dashboard][state] Support task logs from state API

### PR DESCRIPTION
**Pull request by rickyyx**
## Why are these changes needed?
This PR adds support for retrieving task logs from state API with task ID directly. 

In the high level:
- On task execution: it writes magical token including task id + attempt before and after task is run in the worker file.
- On query: it reconstructs the worker file name from querying worker id from the task backend for a task attempt, and look for the magical tokens in the worker file to find beginning and end of the files. 

The end token is used to identify task log in the case of async actor: if async actors product interleaved logs, one task's log will contain all logs from the task and other interleaved logs from other tasks.  (We could probably return some sort of warning to let users know this)

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: dashboard/modules/log/log_agent.py [hunk 1]
```diff
@@ -25,6 +31,37 @@
 DEFAULT_KEEP_ALIVE_INTERVAL_SEC = 1
 
 
+def find_offset_of_content_in_file(
```
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:34]: can you add a unit test?

File: dashboard/modules/log/log_agent.py [hunk 2]
```diff
@@ -252,6 +289,62 @@ async def ListLogs(self, request, context):
...
+        )
+        if end_offset == -1:
+            # No other tasks (might still be running), stream til the end.
+            end_offset = find_end_offset_file(f)
+
+        # Find start
+        task_attempt_start_magic_line = (
+            f"{LOG_PREFIX_TASK_ATTEMPT_START}{task_id}-{attempt_number}\n"
+        )
+        task_attempt_magic_line_offset = find_offset_of_content_in_file(
```
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:327]: how expensive it is to search offset? Should we combine two of them into a single scan?

File: dashboard/modules/log/log_agent.py [hunk 3]
```diff
@@ -279,7 +373,20 @@ async def StreamLog(self, request, context):
...
                 end_offset = find_end_offset_file(f)
 
-                if lines != -1:
+                if task_id is not None:  # Stream from task log.
+                    attempt_number = (
+                        request.attempt_number
+                        if request.HasField("attempt_number")
+                        else 0
+                    )
+                    start_offset, end_offset = await self._find_task_log_offsets(
```
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:382]: Hmm maybe depending on the overhread of file APIs, we can consider offloading them to a different thread not to bother main dashboard ops.

File: dashboard/modules/log/log_manager.py [hunk 4]
```diff
@@ -171,8 +198,48 @@ async def resolve_filename(
...
+            node_id = task.get("node_id", None)
+
+            if worker_id is None or node_id is None:
+                raise FileNotFoundError(
+                    "Could not find log file for task attempt:"
+                    f"{task_id}({attempt_number})."
+                    f"Worker id = {worker_id}, node id = {node_id}"
+                )
+
+            log_filename = await self._resolve_worker_file(
```
Comment (rkooo567) [inline dashboard/modules/log/log_manager.py:233]: Hmm don't we already have the API to find this? I assume we should have one given we already support finding the worker log files

File: python/ray/_raylet.pyx [hunk 5]
```diff
@@ -891,7 +891,16 @@ cdef void execute_task(
                 actor_title = f"{class_name}({args!r}, {kwargs!r})"
                 core_worker.set_actor_title(actor_title.encode("utf-8"))
 
-            worker.record_task_log_start()
+            # Record the task id via magic token in the log file.
+            # This will be used to locate the beginning of logs from a task.
+            attempt_number = core_worker.get_current_task_attempt_number()
```
Comment (rkooo567) [inline python/ray/_raylet.pyx:896]: Note: this may not work with async actor

File: python/ray/tests/test_state_api.py [hunk 6]
```diff
@@ -1719,37 +1701,6 @@ async def test_state_data_source_client_limit_distributed_sources(ray_start_clus
         port = int(node["NodeManagerPort"])
         client.register_raylet_client(node_id, ip, port)
 
-    """
```
Comment (rkooo567) [inline python/ray/tests/test_state_api.py:1722]: why are we removing these tests?

File: python/ray/tests/test_state_api.py [hunk 7]
```diff
@@ -1536,24 +1536,6 @@ async def test_state_data_source_client(ray_start_cluster):
     assert list(result.keys())[0] == job_id
     assert isinstance(result, dict)
 
-    """
```
Comment (rkooo567) [inline python/ray/tests/test_state_api.py:1539]: why are we removing these tests?

**Review by rkooo567 — COMMENTED** [review]
Actually, for async actor, it may not work at all because we don't have unique task ids for each async actor now. I remember you have an "observability only" task id for async actor. Does it work with it?

**Comment by rickyyx** [comment]
> Actually, for async actor, it may not work at all because we don't have unique task ids for each async actor now. I remember you have an "observability only" task id for async actor. Does it work with it?

That was for "parent task" tracking I think. Each async actor task should have the right task id assigned. See tests here https://github.com/ray-project/ray/pull/35101/files#diff-708783b4063f97629661a395ea54466e62f4d0571f4bae1c53a51442308d5109R1141

File: dashboard/modules/log/log_agent.py (hunk 3 above)
Comment (rickyyx) [reply dashboard/modules/log/log_agent.py:382]: yeah, let me do the benchmarking.

File: dashboard/modules/log/log_agent.py (hunk 2 above)
Comment (rickyyx) [reply dashboard/modules/log/log_agent.py:327]: Rearrange to locate start first, and then end.

File: dashboard/modules/log/log_agent.py (hunk 3 above)
Comment (rickyyx) [reply dashboard/modules/log/log_agent.py:382]: Streaming 3GiB data will take 3 seconds. Current approach is offloading the routine into another thread to unblock the event loop.

File: dashboard/modules/log/log_manager.py (hunk 4 above)
Comment (rickyyx) [reply dashboard/modules/log/log_manager.py:233]: ah exactly - that's why I refactored it in the first place.

File: python/ray/tests/test_state_api.py [hunk 8]
```diff
@@ -1536,24 +1536,6 @@ async def test_state_data_source_client(ray_start_cluster):
...
-    Test tasks
```
Comment (rickyyx) [inline python/ray/tests/test_state_api.py:1540]: We no longer query raylets for tasks, this is no longer relevant.

File: python/ray/tests/test_state_api.py [hunk 9]
```diff
@@ -1719,37 +1701,6 @@ async def test_state_data_source_client_limit_distributed_sources(ray_start_clus
...
-        ray.get([long_running.remote() for _ in range(2)])
-
-    # Driver: 2 * f
-    # Each worker: 2 * long_running
-    # -> 2 * f + 4 * long_running
-
-    refs = [f.remote() for _ in range(2)]  # noqa
-
-    async def verify():
-        result = await client.get_task_info(node_id, limit=2)
```
Comment (rickyyx) [inline python/ray/tests/test_state_api.py:1743]: This function is no longer used, it's replaced by `get_all_task_info` which queries the task backend directly.

**Review by rickyyx — COMMENTED** [review]
- Added unit tests for find content in file routine 
- Offload the slow finding find-content routine to a thread pool 
- Addressed comments and revert some unrelated changes.

**Comment by rkooo567** [comment]
> That was for "parent task" tracking I think. Each async actor task should have the right task id assigned. See tests here https://github.com/ray-project/ray/pull/35101/files#diff-708783b4063f97629661a395ea54466e62f4d0571f4bae1c53a51442308d5109R1141

Hmm I see. I guess it is because the random task id is only used inside the async actor method.

File: dashboard/modules/log/log_agent.py (hunk 3 above)
Comment (rkooo567) [reply dashboard/modules/log/log_agent.py:399]: This sgtm

File: dashboard/modules/log/log_agent.py [hunk 10]
```diff
@@ -252,6 +289,79 @@ async def ListLogs(self, request, context):
...
+        """
+
+        # Find start
+        task_attempt_start_magic_line = (
+            f"{LOG_PREFIX_TASK_ATTEMPT_START}{task_id}-{attempt_number}\n"
+        )
+
+        # Offload the heavy IO CPU work to a thread pool to avoid blocking the
+        # event loop for concurrent requests.
+        with concurrent.futures.ThreadPoolExecutor() as pool:
```
Comment (rkooo567) [inline dashboard/modules/log/log_agent.py:320]: You want to specify num_workers here. The default is num_cpus/2 or sth like that which is pretty large IIRC

Comment (rkooo567) [reply dashboard/modules/log/log_agent.py:320]: Also consider maintaing a single pool instead of keep creating a new one?

**Review by rkooo567 — APPROVED** [review]
One last comment regarding the thread pool

**Pull request merged**